### PR TITLE
Auto Query Fixer: Implement ExtractFunctionRange, LocateTableRanges, GetAllKeywords methods in ZetaSQL Helper server

### PR DESCRIPTION
**Pull request by mingen-pan**
Develop ExtractFunctionRange, LocateTableRanges, and GetAllKeywords methods
RPC calls are added
Test cases are included.

### CODE REVIEW

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 1]
```diff
@@ -33,5 +36,51 @@ absl::Status ZetaSqlHelperLocalServiceImpl::Tokenize(
   return absl::OkStatus();
 }
 
+absl::Status ZetaSqlHelperLocalServiceImpl::ExtractFunctionRange(
+    const ExtractFunctionRangeRequest *request,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:40]: why const pointer? if you don't expect to change the request, we should use const reference.

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 2]
```diff
@@ -33,5 +36,51 @@ absl::Status ZetaSqlHelperLocalServiceImpl::Tokenize(
...
+    ExtractFunctionRangeResponse *response) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:41]: nit: ExtractFunctionRangeResponse* response

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 3]
```diff
@@ -33,5 +36,51 @@ absl::Status ZetaSqlHelperLocalServiceImpl::Tokenize(
...
+
+  std::unique_ptr<::bigquery::utils::zetasql_helper::FunctionRange> output;
+  ZETASQL_RETURN_IF_ERROR(
+      ::bigquery::utils::zetasql_helper::ExtractFunctionRange(
+          request->query(), request->line_number(), request->column_number(), &output
+      ));
+
+  auto status_or_function_range_proto = output->ToProto();
+  ZETASQL_RETURN_IF_ERROR(status_or_function_range_proto.status());
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:50]: why return status separately? if error occurs, should it return at line 44? or there could be cases line 44 is ok but getting status() is not ok?

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 4]
```diff
@@ -33,5 +36,51 @@ absl::Status ZetaSqlHelperLocalServiceImpl::Tokenize(
...
+
+  // Convert the proto to heap value and hand over the ownership to the response.
+  auto function_range_proto = new ::bigquery::utils::zetasql_helper::FunctionRangeProto(
+      status_or_function_range_proto.value()
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:54]: you can use unique_ptr and std::move?

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc [hunk 5]
```diff
@@ -33,5 +36,51 @@ absl::Status ZetaSqlHelperLocalServiceImpl::Tokenize(
...
+  );
+  response->set_allocated_function_range(function_range_proto);
+  return absl::OkStatus();
+}
+
+absl::Status ZetaSqlHelperLocalServiceImpl::LocateTableRanges(const LocateTableRangesRequest *request,
+                                                              LocateTableRangesResponse *response) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:61]: pass by reference

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.proto [hunk 6]
```diff
@@ -36,4 +47,33 @@ message TokenizeRequest {
 
 message TokenizeResponse {
   repeated ParseTokenProto parse_token = 1;
+}
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service.proto:50]: Please add comment to the proto below.

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.h [hunk 7]
```diff
@@ -30,6 +30,17 @@ class ZetaSqlHelperLocalServiceGrpcImpl : public ZetaSqlHelperLocalService::Serv
...
                         TokenizeResponse *resp) override;
 
+  grpc::Status ExtractFunctionRange(grpc::ServerContext *context, const ExtractFunctionRangeRequest *request,
+                                    ExtractFunctionRangeResponse *response) override;
+
+  grpc::Status LocateTableRanges(grpc::ServerContext *context,
+                                 const LocateTableRangesRequest *request,
+                                 LocateTableRangesResponse *response) override;
+
+  grpc::Status AllKeywords(grpc::ServerContext *context,
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.h:40]: nit: GetAllKeywordsRequest? method name GetAllKeywords?

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+//
+
+#include "extract_function.h"
+#include "zetasql/public/analyzer.h"
+#include "zetasql/parser/parser.h"
+#include "zetasql_helper/util/util.h"
+
+namespace bigquery::utils::zetasql_helper {
+
+FunctionRange::FunctionRange(const zetasql::ASTFunctionCall *function_call) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:24]: pass by reference if ready only :)

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  proto.set_allocated_function(range_proto);
+
+  // name range transfer
+  auto status_or_name_proto = name.ToProto();
+  ZETASQL_RETURN_IF_ERROR(status_or_name_proto.status());
+
+  range_proto = new zetasql::ParseLocationRangeProto(status_or_name_proto.value());
+  proto.set_allocated_name(range_proto);
+
+  for (auto &argument_range : arguments) {
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:50]: nit: const auto& argument_range: ....

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  std::unique_ptr<zetasql::ParserOutput> parser_output;
+  auto options = BigQueryOptions();
+  auto status = ParseStatement(query, options.GetParserOptions(), &parser_output);
+  if (!status.ok()) {
+    return status;
+  }
+
+  auto offset = get_offset(query, row, column);
+  auto predicator = [offset](const zetasql::ASTNode *node) {
+    return node->GetParseLocationRange().start().GetByteOffset() == offset &&
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:75]: nit: zetasql::ASTNode* node

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        node->node_kind() == zetasql::ASTNodeKind::AST_FUNCTION_CALL;
+  };
+
+  auto candidate = find_node(parser_output->statement(), predicator);
+  if (candidate == nullptr) {
+    return absl::Status(absl::StatusCode::kInvalidArgument, "Line and/or column numbers are incorrect");
+  }
+
+  auto function_call = dynamic_cast<const zetasql::ASTFunctionCall *>(candidate);
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:84]: nit: ASTFunctionCall*

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.h [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+// A struct to store the range of different components of a function. It includes the range
+// of the whole function, the name, and each arguments.
+struct FunctionRange {
+  zetasql::ParseLocationRange function;
+  zetasql::ParseLocationRange name;
+  std::vector<zetasql::ParseLocationRange> arguments;
+
+  FunctionRange(const zetasql::ASTFunctionCall *function_call);
+
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/extract_function.h:37]: nit:

File: tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+TEST_F(LocationTest, LocateTableTest2) {
+  std::string query = "Select max(foo) from bigquery-public-data.mock.survey_2017 group by bar limit 10";
+  std::string table_regex = "bigquery-public-data\\.mock\\.survey_2017";
+
+  std::vector<zetasql::ParseLocationRange> ranges;
+  auto status = LocateTableRanges(query, table_regex, ranges);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(1, ranges.size());
+//  EXPECT_EQ(2, ranges[0].start().GetByteOffset())
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc:58]: nit: delete the line?

File: tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+      "where status = '`bigquery-public-data.austin_311.311_request`'";
+
+  std::string table_regex = "(bigquery-public-data.)?austin_311.311_request";
+
+  std::vector<zetasql::ParseLocationRange> ranges;
+  auto status = LocateTableRanges(query, table_regex, ranges);
+
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(2, ranges.size());
+//  EXPECT_EQ(2, ranges[0].start().GetByteOffset())
```
Comment (kikkyo) [inline tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc:40]: nit: delete the line?

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 1 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:40]: Good point. Changed to reference.

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 2 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:41]: Done

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 3 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:50]: Yes. Line 44 makes sure the ExtractFunctionRange is correct. Then we get a FunctionRange object. Then we need to serialize it to proto, which is line 50. Here may  occur errors.

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 4 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:54]: I think this is how proto specifies the ownership - assign a pointer to its field. The child pointer will be released automatically when the owner is released. We don't need the unique_ptr here.

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc (hunk 9 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:50]: Just realized the formatter will concat the & and * with the variable..

File: tools/zetasql_helper/zetasql_helper/local_service/local_service.cc (hunk 5 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service.cc:61]: Done

File: tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.h (hunk 7 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/local_service/local_service_grpc.h]: Yes, I have changed them all to `GetAllKeywords`

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc (hunk 8 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc]: Done

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc (hunk 11 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc]: Done

File: tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc (hunk 10 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/extract_function.cc:68]: Done

File: tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc (hunk 13 above)
Comment (mingen-pan) [reply tools/zetasql_helper/zetasql_helper/scanner/scanner_test.cc:60]: I have changed it to two new EXPECT_EQ

**Review by kikkyo — APPROVED** [review]

**Pull request merged**